### PR TITLE
Parse ASSUME declarations names

### LINKED ISSUE #1835: Names of assumptions with `ASSUME` are not parsed into the IR
## Description

From all possible TLA+ declarations Apalache parses, `ASSUME` is the only one in which the name is ignored, even tho it is always present in TLA+.

Even if it's not relevant for Apalache, I believe we should parse it and represent it on the IR so it is more consistent. I know this IR is not a public interface for anyone to use yet, but I'm writing a parser for it and would not like to have a special case for handling this silly case.

## Input specification

```tla
---------------------------- MODULE AssumeExample----------------------------
CONSTANT N

ASSUME myName == N = TRUE
=============================================================================
```

## The command line parameters used to run the tool

```
apalache-mc parse --output=AssumeExample.json AssumeExample.tla
```

## Expected behavior

At `AssumeExample.json`:

```
{
  "source": {
    "filename": "AssumeExample",
    "from": {
      "line": 4,
      "column": 1
    },
    "to": {
      "line": 4,
      "column": 25
    }
  },
  "type": "Untyped",
  "kind": "TlaAssumeDecl",
  "name": "myName", // <-------- this line is currently missing
  "body": { ... }
}
```

## System information

- Apalache version [`apalache-mc version`]: 0.25.6

### PR DESCRIPTION
**Pull request by fan-tom**
Closes #1835

- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [x] [Entries added to `./unreleased/`][changelog format] for any new functionality

[changelog format]: https://github.com/informalsystems/apalache/blob/main/CONTRIBUTING.md#how-to-record-a-change

### CODE REVIEW

File: tla-io/src/main/scala/at/forsyte/apalache/io/json/TlaToJson.scala [hunk 1]
```diff
@@ -188,13 +188,20 @@ class TlaToJson[T <: JsonRepresentation](
...
-        withLoc(
+        val fields = Array[(String, T)](
             typeFieldName -> typeTagPrinter(decl.typeTag),
             kindFieldName -> "TlaAssumeDecl",
             "body" -> bodyJson,
         )
+
+        val f = definedName match {
+          case None       => fields
+          case Some(name) => fields :+ ("name" -> name: (String, T))
```
Comment (fan-tom) [inline tla-io/src/main/scala/at/forsyte/apalache/io/json/TlaToJson.scala:201]: Is it critical that the name will appear after the body in the json?

File: tla-io/src/test/scala/at/forsyte/apalache/io/TestPrettyWriterWithTypes.scala [hunk 2]
```diff
@@ -97,4 +97,32 @@ class TestPrettyWriterWithTypes extends AnyFunSuite with BeforeAndAfterEach {
         |""".stripMargin
     assert(expected == stringWriter.toString)
   }
+
+  test("unnamed assume declaration") {
+    val decl = TlaAssumeDecl(None, tla.eql(tla.name("x"), tla.bool(true)))
```
Comment (fan-tom) [inline tla-io/src/test/scala/at/forsyte/apalache/io/TestPrettyWriterWithTypes.scala:102]: Should I assign any `typeTag` here?

File: tla-typechecker/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala [hunk 3]
```diff
@@ -45,7 +45,12 @@ class ToEtcExpr(
         val operType = OperT1(Seq(BoolT1), BoolT1)
         val application = mkUniqApp(Seq(operType), this(d.body))
         // We have to introduce a lambda abstraction, as the type checker is expecting this form.
-        mkLet(BlameRef(d.ID), "__Assume_" + d.ID, mkAbs(ExactRef(d.ID), application), inScopeEx)
+        mkLet(
+            BlameRef(d.ID),
+            "__Assume_" + d.definedName.getOrElse(d.ID.toString),
```
Comment (fan-tom) [inline tla-typechecker/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala:50]: Should we try to utilize the defined name here, or leave it as is?

File: tlair/src/test/scala/at/forsyte/apalache/tla/lir/IrGenerators.scala [hunk 4]
```diff
@@ -348,7 +348,7 @@ trait IrGenerators extends TlaType1Gen {
     for {
       ex <- exGen
       tt <- genTypeTag
-    } yield TlaAssumeDecl(ex).withTag(tt)
+    } yield TlaAssumeDecl(None, ex).withTag(tt)
```
Comment (fan-tom) [inline tlair/src/test/scala/at/forsyte/apalache/tla/lir/IrGenerators.scala:351]: Should we generate named version?

File: .unreleased/features/01.md
Comment (fan-tom) [inline .unreleased/features/01.md:1]: Should I record it in the changelog?

File: tla-io/src/main/scala/at/forsyte/apalache/io/json/TlaToJson.scala (hunk 1 above)
Comment (shonfeder) [reply tla-io/src/main/scala/at/forsyte/apalache/io/json/TlaToJson.scala:201]: No!

**Comment by fan-tom** [comment]
I added additional tests to increase code coverage based on the Codecov report, but these tests look ad-hoc, @shonfeder  please advise whether I should preserve them or replace with something else, or just drop and ignore Codecov warnings

File: tla-io/src/test/scala/at/forsyte/apalache/io/quint/TestQuintEx.scala [hunk 5]
```diff
@@ -789,4 +793,18 @@ class TestQuintEx extends AnyFunSuite {
...
+    val nullaryOps = Set[String]()
+
+    val name = "myAssume"
+    val namedAssume = QuintAssume(1, name, QuintBool(2, true))
+    val tlaNamedAssumeDef = translator.tlaDef(namedAssume).run(nullaryOps).get._2
+    assert(tlaNamedAssumeDef == TlaAssumeDecl(Some(name), ValEx(TlaBool(true))(Typed(BoolT1)))(Typed(BoolT1)))
+
+    val unnamedAssume = QuintAssume(1, "_", QuintBool(2, true))
+    val tlaUnnamedAssumeDef = translator.tlaDef(unnamedAssume).run(nullaryOps).get._2
+    assert(tlaUnnamedAssumeDef == TlaAssumeDecl(None, ValEx(TlaBool(true))(Typed(BoolT1)))(Typed(BoolT1)))
```
Comment (shonfeder) [inline tla-io/src/test/scala/at/forsyte/apalache/io/quint/TestQuintEx.scala:808]: Sorry this is so gnarly to test!

File: tla-typechecker/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala (hunk 3 above)
Comment (shonfeder) [reply tla-typechecker/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala:50]: Let's use the name attribute!

```suggestion
            d.name,
```

File: tla-typechecker/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala [hunk 6]
```diff
@@ -131,7 +136,7 @@ class ToEtcExpr(
     OperT1(nBools, BoolT1)
   }
 
-  // Valid when the input seq has two items, the first of which is a VlaEx(TlaStr(_))
+  // Valid when the input seq has two items, the first of which is a ValEx(TlaStr(_))
```
Comment (shonfeder) [inline tla-typechecker/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala:139]: Thanks :)

File: tla-typechecker/src/test/scala/at/forsyte/apalache/tla/typecheck/etc/TestToEtcExprDecls.scala [hunk 7]
```diff
@@ -116,7 +116,8 @@ class TestToEtcExprDecls extends AnyFunSuite with ToEtcExprBase with BeforeAndAf
     // The body is wrapped with the application of an operator that has the signature Bool => Bool.
     // This allows us to check that the assumption has Boolean type.
     val application = mkUniqApp(Seq(parser("Bool => Bool")), assumption)
-    val expected = mkUniqLet("__Assume_" + assume.ID, mkUniqAbs(application), terminal)
+    val expected =
+      mkUniqLet("__Assume_" + assume.definedName.getOrElse(assume.ID.toString), mkUniqAbs(application), terminal)
```
Comment (shonfeder) [inline tla-typechecker/src/test/scala/at/forsyte/apalache/tla/typecheck/etc/TestToEtcExprDecls.scala:120]: To match the change in `ToEtcExpr.scala`

```suggestion
      mkUniqLet(assume.name, mkUniqAbs(application), terminal)
```

File: tlair/src/test/scala/at/forsyte/apalache/tla/lir/TestPrinter.scala [hunk 8]
```diff
@@ -195,6 +195,14 @@ class TestPrinter extends AnyFunSuite with TestingPredefs {
...
   }
 
+  test("Test UTF8: TlaAssumeDecl") {
+    val namedAssume: String = toUtf(TlaAssumeDecl(Some("myAssume"), tla.eql(tla.name("x"), tla.bool(true))))
+    assert(namedAssume == s"ASSUME myAssume ≜ x = TRUE")
+
+    val unnamedAssume: String = toUtf(TlaAssumeDecl(None, tla.eql(tla.name("x"), tla.bool(true))))
+    assert(unnamedAssume == s"ASSUME x = TRUE")
+  }
+
```
Comment (shonfeder) [inline tlair/src/test/scala/at/forsyte/apalache/tla/lir/TestPrinter.scala:205]: Awesome!

**Review by shonfeder — COMMENTED** [review]
Awesome work! It was a pleasure to review. Thanks so much for picking this up :)

I have a few small change requests, and a suggestion or two, but they are all minor nits.

**Comment by shonfeder** [comment]
> I added additional tests to increase code coverage based on the Codecov report, but these tests look ad-hoc, @shonfeder please advise whether I should preserve them or replace with something else, or just drop and ignore Codecov warnings

Those I think the tests look OK to me :) In the future, if you don't think a test is meaningful, no need to add it just to get the coverage metrics, but since they are already here, let's keep 'em.

File: tlair/src/test/scala/at/forsyte/apalache/tla/lir/IrGenerators.scala (hunk 4 above)
Comment (shonfeder) [reply tlair/src/test/scala/at/forsyte/apalache/tla/lir/IrGenerators.scala:351]: Ah, let's not worry about it for now. We can allow add that in a follow up. The generators are do for an overhaul at some point any how.

**Review by bugarela — APPROVED** [review]
Thanks for taking care of this!

**Pull request merged**